infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntrySinkBufferManager.java

```diff
@@ -74,4 +75,24 @@ public boolean verifyMessageType(LogReplicationEntry entry) {
 
         return true;
     }
+
+    public void processBuffer() {
+        /**
+         *  For each message in the  buffer, if its timestamp is smaller than last processed log entry's timestamp,
+         *  skip processing and remove it from buffer.
+         *  If its preTs and currentTs is overlapping with the last processed log entry's timestamp, process it.
+         */
+        for (LogReplicationEntry entry : buffer.values()) {
+            LogReplicationEntryMetadata metadata = entry.getMetadata();
+            if (metadata.getTimestamp() < lastProcessedSeq) {
+                //remove it
+                buffer.remove(metadata.getPreviousTimestamp());
+            } else if (metadata.getPreviousTimestamp() <= lastProcessedSeq && metadata.getTimestamp() > lastProcessedSeq) {
+                sinkManager.processMessage(entry);
+                ackCnt++;
+                buffer.remove(lastProcessedSeq);
```

Comment:
For each message, it has the previous time stamp and the currentTimestamp( the max ts of all transactions in this message). So if all transactions in one message have been processed, we will remove it. If it has transactions that haven't been processed, we will still buffer it and process it later.

This happens because the possible leadership change and the two leaders at the active cluster may packaging the messages differently as the polling transaction log speed  are different at the different nodes.